### PR TITLE
Feature/serviceusage constraint

### PR DESCRIPTION
**Pull request by reechar-goog**

### CODE REVIEW

File: validator/test/fixtures/constraints/serviceusage_disallow_cloudvisionapi/data.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+apiVersion: constraints.gatekeeper.sh/v1alpha1
+kind: GCPServiceUsageConstraintV1
+metadata:
+  name: disallow_cloudvision
+  # annotations:
+  #   description: Prevent public users from having access to resources via IAM
```
Comment (morgante) [inline validator/test/fixtures/constraints/serviceusage_disallow_cloudvisionapi/data.yaml]: This description looks incorrect, can we drop?

File: validator/test/fixtures/constraints/serviceusage_allow_compute/data.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+apiVersion: constraints.gatekeeper.sh/v1alpha1
+kind: GCPServiceUsageConstraintV1
+metadata:
+  name: allow_only_computeapi
+  # annotations:
+  #   description: Prevent public users from having access to resources via IAM
```
Comment (morgante) [inline validator/test/fixtures/constraints/serviceusage_allow_compute/data.yaml]: Description is out of date.

File: validator/test/fixtures/constraints/serviceusage_disallow_cloudvisionapi/data.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+spec:
+  severity: high
+  match:
+    target: ["organization/*"]
+    exclude: [] # optional, default is no exclusions
+  parameters:
+    mode: disallowed
```
Comment (morgante) [inline validator/test/fixtures/constraints/serviceusage_disallow_cloudvisionapi/data.yaml]: Nit. `denied` is probably more consistent.

File: policies/templates/gcp_serviceusage_allowed_services_v1.yaml [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: templates.gatekeeper.sh/v1alpha1
+kind: ConstraintTemplate
+metadata:
+  name: gcp-serviceusage_allowed_services_v1
```
Comment (morgante) [inline policies/templates/gcp_serviceusage_allowed_services_v1.yaml]:
```suggestion
  name: gcp-serviceusage-allowed-services-v1
```

File: validator/serviceusage_service.rego [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+	service_usage := asset.resource.data
+	parent := service_usage.parent
+	state := service_usage.state
+	service := extract_service(service_usage.name)
+
+	mode := lib.get_default(params, "mode", "allow")
+
+	state == "ENABLED"
+	matches_found = [m | m = params.services[_]; glob.match(m, [], service)]
```
Comment (morgante) [inline validator/serviceusage_service.rego]: Does this need to be a glob or should it be a literal match?

Comment (reechar-goog) [reply validator/serviceusage_service.rego]: @morgante are you suggesting to use: re_match ? I think we probably want a literal match, but don't understand the side effects of using glob

Comment (morgante) [reply validator/serviceusage_service.rego]: I mean just `m == service`.

Comment (reechar-goog) [reply validator/serviceusage_service.rego]: ah yea that makes sense, I don't really see regex being useful for googleapi services

**Pull request merged**